rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,669 @@
+
+Because a Rust pointer is not subjected to borrow checking and has no obligation
+of pointing towards a valid memory location, this means that using a Volatile
+wrapper in any way is unsafe.
+
+As a second difference, in addition to familiar atomic memory operations,
+volatile types expose the `load_not_atomic()` and `store_not_atomic()` methods.
+As their name suggest, these memory operations are not considered to be atomic
+by the compiler, and are therefore not safe to concurrently invoke in multiple
+threads.
+
+On hardware with global cache coherence, a category which encompasses the vast
+majority of Rust's supported compilation targets, use of these methods will
+generate exactly the same code as using the `load()` and `store()` atomic access
+methods with `Relaxed` memory ordering, with the only difference being that
+data races are Undefined Behavior from the compiler's point of view. When that
+is the case, safer `Relaxed` atomic volatile operations should be preferred to
+their non-atomic equivalents.
+
+Unfortunately, however, not all of Rust's compilation targets exhibit global
+cache coherence. GPU hardware, such as the `nvptx` target, may only exhibit
+cache coherence among local "blocks" of threads. And abstract machines like WASM
+may not guarantee cache coherence at all without specific precautions. On those
+compilation targets, `Relaxed` loads and stores may either be unavailable, or
+lead to the generation of machine instructions more complex than native loads
+and stores, which may not be wanted where maximal hardware control or CPU
+performance is desired.
+
+It is only for the sake of providing an alternative to the current
+`ptr::[read|write]_volatile` mechanism on such platforms that the
+`[load|store]_not_atomic()` functions are being proposed, and they should not
+be used where better alternative exists.
+
+Finally, unlike with atomics, the compiler is not allowed to optimize the above
+function into the following...
+
+```rust
+unsafe fn do_volatile_things(target: NonNull<VolatileU8>) -> u8 {
+    target.store(42, Ordering::Relaxed);
+    42
+}
+```
+
+...or even the following, which it could normally do if it the optimizer managed
+to prove that no other thread has access to the `VolatileU8` variable:
+
+```rust
+unsafe fn do_volatile_things(_target: NonNull<VolatileU8>) -> u8 {
+    42
+}
+```
+
+This is the definining characteristic of volatile operations, which makes them
+suitable for sensitive memory manipulations such as cryptographic secret erasure
+or memory-mapped I/O.
+
+---
+
+Experienced Rust users may be familiar with the previously existing
+`std::ptr::read_volatile()` and `std::ptr::write_volatile()` functions, or
+equivalent methods of raw pointer objects, and wonder how the new facilities
+provided by `std::volatile` compare to those methods.
+
+The answer is that this new volatile data access API supersedes its predecessor,
+which is now _deprecated_, by improving upon it in several ways:
```

Comment:
I am not a fan of deprecating the old methods. There are cases where you do not actually care whether the accesses tear, e.g. cryptography and (at least some instances of) shared memory.

Rust libstd should still provide methods for arbitrarily-sized volatile accesses, it should just be clearly documented that they tear (and possibly we could even document *how* they tear).

Reply:
I think that we should separate the deprecation issue from adding the new APIs. There are arguments that can be made in favour of deprecating and keeping the already-existing APIs, but we don't really need to make these arguments here. If someone wants to submit an RFC to deprecate the old methods, they can make their case in a separate RFC.

Reply:
Agreed, will do in next round of edits.

Reply:
Done.